### PR TITLE
Fix: Schema validation for the correct definition of n..m relationships - must be non-nullable lists and entries

### PR DESCRIPTION
**Pull request by danstarns**
# Description

Enforces the validation stated in this RFC : [https://github.com/neo4j/graphql/blob/dev/docs/rfcs/rfc-003-relationship-cardinality.md#invalid-relationship-definition](https://github.com/neo4j/graphql/blob/dev/docs/rfcs/rfc-003-relationship-cardinality.md#invalid-relationship-definition)

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-graphql-docs-736.surge.sh

File: packages/graphql/tests/tck/tck-test-files/directives/node/node-label-union.test.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ import { Neo4jGraphQL } from "../../../../../src";
 import { createJwtRequest } from "../../../../utils/create-jwt-request";
 import { formatCypher, translateQuery, formatParams } from "../../../utils/tck-test-utils";
 
-describe("Node directive with unions", () => {
+describe("!", () => {
```
Comment (angrykoala) [inline packages/graphql/tests/tck/tck-test-files/directives/node/node-label-union.test.ts:26]: erm, I guess this is a typo :sweat_smile:

**Review by angrykoala — CHANGES_REQUESTED** [review]

File: packages/graphql/tests/tck/tck-test-files/directives/node/node-label-union.test.ts (hunk 1 above)
Comment (danstarns) [reply packages/graphql/tests/tck/tck-test-files/directives/node/node-label-union.test.ts:26]: Good catch!

File: packages/graphql/tests/integration/update.int.test.ts [hunk 2]
```diff
@@ -1015,9 +1015,6 @@ describe("update", () => {
                     id
                     photos {
                         id
-                        color {
```
[the remark this answers is not in the document]
Comment (danstarns) [reply packages/graphql/tests/integration/update.int.test.ts]: Great catch!

File: docs/modules/ROOT/pages/ogm/selection-set.adoc [hunk 3]
```diff
@@ -39,7 +39,7 @@ When using the OGM, you do not have to provide a selection set by default. Doing
 type Movie {
     id: ID
     name: String
-    genres: [Genre] @relationship(type: "IN_GENRE", direction: OUT)
+    genres: [Genre]! @relationship(type: "IN_GENRE", direction: OUT)
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/ogm/selection-set.adoc:42]: The definitions in here aren't valid right? Should be:

```suggestion
    genres: [Genre!]! @relationship(type: "IN_GENRE", direction: OUT)
```

Same for the other two changes in this file.

File: packages/graphql/tests/tck/tck-test-files/directives/auth/arguments/where/where.test.ts [hunk 4]
```diff
@@ -35,14 +35,14 @@ describe("Cypher Auth Where", () => {
...
-                content: [Search] @relationship(type: "HAS_POST", direction: OUT) # something to test unions
+                posts: [Post!]! @relationship(type: "HAS_POST", direction: OUT)
+                content: [Search!]! @relationship(type: "HAS_POST", direction: OUT) # something to test unions
             }
 
             type Post {
                 id: ID
                 content: String
-                creator: User @relationship(type: "HAS_POST", direction: IN)
+                creator: User! @relationship(type: "HAS_POST", direction: IN)
```
Comment (darrellwarde) [inline packages/graphql/tests/tck/tck-test-files/directives/auth/arguments/where/where.test.ts:45]: I've noticed that you've changed all relationships, even single ones, to be non-nullable. Just to confirm, the following wouldn't throw and error right?

```graphql
creator: User @relationship(type: "HAS_POST", direction: IN)
```

File: packages/graphql/tests/schema/rfcs/rfc-003.test.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { gql } from "apollo-server";
+import { GraphQLSchema } from "graphql";
+import { Neo4jGraphQL } from "../../../src";
+
+describe("schema/rfs/003", () => {
+    const msg = "Cannot represent null for the relationship Source.targets, try using NonNull operator '!'";
```
Comment (darrellwarde) [inline packages/graphql/tests/schema/rfcs/rfc-003.test.ts:26]: Could the thrown message possibly be more specific and guiding, for example:

"List type relationship fields must be non-nullable and have non-nullable entries, please change type of `Source.targets` to `[Target!]!`"

We could possibly also add a link to the documentation in this message for background, which leads to me my next question - has this been documented? Perhaps it should be explained in https://neo4j.com/docs/graphql-manual/current/type-definitions/relationships/?

File: packages/graphql/tests/integration/rfcs/rfc-003.test.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
+/*
```
Comment (darrellwarde) [inline packages/graphql/tests/integration/rfcs/rfc-003.test.ts:1]: None of the code for the tests in this file is actually implemented yet? Especially the unique field stuff if we're moving this out of scope? Has this file been committed accidentally?

**Review by darrellwarde — CHANGES_REQUESTED** [review]

File: packages/graphql/tests/tck/tck-test-files/directives/auth/arguments/where/where.test.ts (hunk 4 above)
Comment (darrellwarde) [reply packages/graphql/tests/tck/tck-test-files/directives/auth/arguments/where/where.test.ts:45]: I wrote this comment early in my review - I now see that the example I gave would be perfectly valid, I just found it interesting how you've replaced all nullable fields with non-nullable fields in the tests. 🙂

File: docs/modules/ROOT/pages/ogm/selection-set.adoc (hunk 3 above)
Comment (danstarns) [reply docs/modules/ROOT/pages/ogm/selection-set.adoc]: Great observation this one slipped under the radar.

File: packages/graphql/tests/tck/tck-test-files/directives/auth/arguments/where/where.test.ts (hunk 4 above)
Comment (danstarns) [reply packages/graphql/tests/tck/tck-test-files/directives/auth/arguments/where/where.test.ts:45]: True I believe I have replaced if not most nullable fields with non-nullable fields, and this was probably the find and replace regex I used.

File: packages/graphql/tests/integration/rfcs/rfc-003.test.ts (hunk 6 above)
Comment (danstarns) [reply packages/graphql/tests/integration/rfcs/rfc-003.test.ts:1]: Yeah, it has! Thanks for pointing it out!

File: packages/graphql/tests/schema/rfcs/rfc-003.test.ts (hunk 5 above)
Comment (danstarns) [reply packages/graphql/tests/schema/rfcs/rfc-003.test.ts]: Sure!

File: docs/modules/ROOT/pages/ogm/selection-set.adoc [hunk 7]
```diff
@@ -78,7 +78,7 @@ const typeDefs = `
     type Movie {
         id: ID
         name: String
-        genres: [Genre] @relationship(type: "IN_GENRE", direction: OUT)
+        genres: [Genre]! @relationship(type: "IN_GENRE", direction: OUT)
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/ogm/selection-set.adoc:81]:
```suggestion
        genres: [Genre!]! @relationship(type: "IN_GENRE", direction: OUT)
```

File: docs/modules/ROOT/pages/ogm/selection-set.adoc [hunk 8]
```diff
@@ -124,7 +124,7 @@ const typeDefs = `
...
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/ogm/selection-set.adoc:127]:
```suggestion
        genres: [Genre!]! @relationship(type: "IN_GENRE", direction: OUT)
```

File: docs/modules/ROOT/pages/type-definitions/relationships.adoc [hunk 9]
```diff
@@ -240,3 +240,21 @@ query {
...
+    name: String!
+    posts: [Post!]! @relationship(type: "HAS_POST", direction: OUT)
+}
+
+type Post {
+    name: String!
+}
+----
+
+the relationship at `User.posts` is considered a many relationship. Relationships such as the one above should always be of type NonNullListType and NonNullNamedType meaning both the array and the type inside of it should have a `!`.
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/type-definitions/relationships.adoc:260]:
```suggestion
The relationship at `User.posts` is considered a "many" relationship. Relationships such as the one above should always be of type `NonNullListType` and `NonNullNamedType`, meaning both the array and the type inside of it should have a `!`.
```

File: docs/modules/ROOT/pages/type-definitions/relationships.adoc [hunk 10]
```diff
@@ -240,3 +240,21 @@ query {
     }
 }
 ----
+
+== Cardinality
+
+Neo4j GraphQL enforces that relationships defined as 'many' so given this example:
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/type-definitions/relationships.adoc:246]:
```suggestion
The Neo4j GraphQL Library has type definition requirements for "many" relationships, so given this example: 
```

**Review by darrellwarde — CHANGES_REQUESTED** [review]
Just a couple more change requests (all as suggestions so can just be applied from here), and then happy to approve following that! 😄

**Review by darrellwarde — APPROVED** [review]
Looks great! Just changed the title to something more descriptive for the changelog. 🙂

**Pull request merged**